File: src/lib/strategyStore.ts
/**
 * Frontend data-access layer for the unified `.rhai` strategy store
 * (AGT-651): the directory of `.rhai` files at `~/.wickd/strategies/` that
 * the wickd CLI and this app share. Read-only on purpose — the app is a
 * strategy viewer/runner; authoring happens through the CLI
 * (`wickd strategy add/update/remove`, see docs/strategy-store.md).
 *
 * Deliberately a separate module from `localStore.ts`: the local store is
 * the SQLite database (`~/.wickd/app.db`), this is the filesystem store.
 */

import { invoke } from '@tauri-apps/api/core';
import type { ParameterDefinition, IndicatorDefinition } from '../types/strategy';

/** One `.rhai` strategy in the store (metadata parsed from script comments). */
export interface StoreStrategy {
  /** Canonical name = the file stem (`revert_adx.rhai` -> `revert_adx`). */
  name: string;
  /** Absolute path of the `.rhai` file. */
  path: string;
  /** Whether the script passes wickd-core's validate_script. */
  valid: boolean;
  /** Validation error when `valid` is false. */
  error?: string;
  /** Parameters declared in the script's `@parameters` metadata. */
  parameters: ParameterDefinition[];
  /** Indicators declared in the script's `@indicators` metadata. */
  indicators: IndicatorDefinition[];
  /** File modification time (epoch ms). */
  modified_at: number;
  /** Stable content fingerprint. */
  content_hash: string;
}

/** `store_read_strategy` result: the entry plus its full source. */
export interface StoreStrategyWithSource extends StoreStrategy {
  source: string;
}

/** Prefix marking in-memory Strategy rows that come from the file store. */
export const STORE_STRATEGY_ID_PREFIX = 'rhai:';

/** Whether a strategy id refers to a (read-only) file-store entry. */
export function isStoreStrategy(id: string): boolean {
  return id.startsWith(STORE_STRATEGY_ID_PREFIX);
}

/** The store id for a `.rhai` strategy name. */
export function storeStrategyId(name: string): string {
  return `${STORE_STRATEGY_ID_PREFIX}${name}`;
}

/** List every `.rhai` strategy in `~/.wickd/strategies/`. */
export function storeListStrategies(): Promise<StoreStrategy[]> {
  return invoke<StoreStrategy[]>('store_list_strategies');
}

/** Read one stored strategy (metadata + full source) by name. */
export function storeReadStrategy(name: string): Promise<StoreStrategyWithSource> {
  return invoke<StoreStrategyWithSource>('store_read_strategy', { name });
}

File: src-tauri/src/commands/strategy_store.rs
//! Tauri commands over the unified `.rhai` strategy store
//! (`~/.wickd/strategies/`, AGT-651).
//!
//! Read-only on purpose: the desktop app is a strategy **viewer/runner**
//! (list, inspect source, run backtests). Authoring goes through the wickd
//! CLI (`wickd strategy add/update/remove` — see `docs/strategy-store.md`),
//! so there are no write commands here and the app can never diverge from
//! the CLI's view of the store.

use serde::Serialize;
use wickd_core::strategy_store::{StoredStrategy, StrategyStore};

/// A stored strategy plus its source, for the read-only source viewer.
#[derive(Debug, Serialize)]
pub struct StoredStrategyWithSource {
    #[serde(flatten)]
    pub entry: StoredStrategy,
    pub source: String,
}

/// List every `.rhai` strategy in the store with parsed metadata. A missing
/// store directory is an empty list, never an error.
#[tauri::command]
pub fn store_list_strategies() -> Result<Vec<StoredStrategy>, String> {
    StrategyStore::open_default()?.list()
}

/// Read one stored strategy (metadata + full source) by name.
#[tauri::command]
pub fn store_read_strategy(name: String) -> Result<StoredStrategyWithSource, String> {
    let store = StrategyStore::open_default()?;
    let (entry, source) = store
        .read(&name)?
        .ok_or_else(|| format!("no stored strategy '{name}'"))?;
    Ok(StoredStrategyWithSource { entry, source })
}
